crates/extensions/c8y_mapper_ext/src/alarm_converter.rs

```diff
@@ -33,62 +34,67 @@ impl AlarmConverter {
 
     pub(crate) fn try_convert_alarm(
         &mut self,
+        source: &EntityMetadata,
         input_message: &Message,
+        alarm_type: &str,
     ) -> Result<Vec<Message>, ConversionError> {
         let mut output_messages: Vec<Message> = Vec::new();
         match self {
             Self::Syncing {
                 pending_alarms_map,
                 old_alarms_map: _,
             } => {
-                let alarm_id = input_message
-                    .topic
-                    .name
-                    .strip_prefix(TEDGE_ALARMS_TOPIC)
-                    .expect("Expected tedge/alarms prefix")
-                    .to_string();
+                let alarm_id = input_message.topic.name.to_string();
                 pending_alarms_map.insert(alarm_id, input_message.clone());
             }
             Self::Synced => {
-                //Regular conversion phase
+                // Regular conversion phase
                 let mqtt_topic = input_message.topic.name.clone();
                 let mqtt_payload = input_message.payload_str().map_err(|e| {
-                    ThinEdgeJsonDeserializerError::FailedToParsePayloadToString {
+                    ThinEdgeAlarmDeserializerError::FailedToParsePayloadToString {
                         topic: mqtt_topic.clone(),
                         error: e.to_string(),
                     }
                 })?;
 
-                let tedge_alarm =
-                    ThinEdgeAlarm::try_from(&mqtt_topic, mqtt_payload).map_err(|e| {
-                        ThinEdgeJsonDeserializerError::FailedToParseJsonPayload {
-                            topic: mqtt_topic,
-                            error: e.to_string(),
-                            payload: mqtt_payload.chars().take(50).collect(),
-                        }
-                    })?;
-                let c8y_alarm = C8yCreateAlarm::try_from(&tedge_alarm)?;
+                let tedge_alarm = if mqtt_payload.is_empty() {
+                    // Clearing an alarm
+                    ThinEdgeAlarm::new(alarm_type, source)
+                } else {
+                    // Creating an alarm
+                    let tedge_alarm_data =
```

Comment:
I'd just restore that old `ThinEdgeAlarm::try_from` function or something like `impl TryFrom<&Message> for ThinEdgeAlarm`

Reply:
Just not that `impl TryFrom<&Message> for ThinEdgeAlarm` is not ideal as `alarm_type` and `source` are given by the upper level. Re-invention of Message to `alarm_type` and `source` are not smart.

Here we need 3 things to create `ThinEdgeAlarm`.
1. `alarm_type` (originally from topic, but here derived from upper level)
2. `source` (originally from topic, but here derived from upper level)
3. `mqtt_payload` (get from `Message`).

Technically we can create a generic struct and implement `TryFrom` for the struct as these 3 parameters are the same for measurements and events. However, I'd address this level of sophistication later.

In short, I will create a function which takes the 3 fields and return `Result<ThinEdgeAlarm, Err>` in this PR.

Reply:
Done with ec55b7328b99294efd1e2e16b7430bb7bed5a371